### PR TITLE
Feature/ummeli allow user to filter by city

### PR DESCRIPTION
**Pull request by miltontony**
Mostly an integration of Django Atlas and how it will work with our current Ummeli stuff

### CODE REVIEW

File: ummeli/opportunities/models.py [hunk 1]
```diff
@@ -28,6 +29,15 @@ class Province(models.Model):
     def __unicode__(self):  # pragma: no cover
         return '%s' % dict(PROVINCE_CHOICES)[self.province]
 
+    @classmethod
+    def from_str(cls, str):
+        result = [i for i, p in PROVINCE_CHOICES\
+                    if re.sub('[\s-]', '', p.lower()) ==\
+                        re.sub('[\s-]', '', str.lower())]
```
Comment (smn) [inline ummeli/opportunities/models.py:38]: You don't need to use the `\` in list comprehensions.

File: ummeli/opportunities/models.py [hunk 2]
```diff
@@ -28,6 +29,15 @@ class Province(models.Model):
...
+        if any(result):
+            return cls.objects.get(pk=result[0])
```
Comment (smn) [inline ummeli/opportunities/models.py:44]: What if there are more than 1 results? Are you ok with losing one?

File: ummeli/opportunities/views.py [hunk 3]
```diff
@@ -58,6 +58,14 @@ def get_queryset(self):
 
         position = self.request.session['location']['position']
 
+        if self.request.session.get('override_location'):
+            tasks = MicroTask.permitted.filter(campaign__pk=campaign.pk)\
+                                        .order_by('province', 'location__city')
+            province = self.request.session['province']
+            if province > 0:
```
Comment (smn) [inline ummeli/opportunities/views.py]: What's the zero here? Please use a constant or something more descriptive that'll help us to understand this code 9 months from now.

File: ummeli/opportunities/views.py [hunk 4]
```diff
@@ -197,3 +219,20 @@ def task_instructions(request, slug):
...
     return render(request, 'opportunities/microtasks/microtask_instructions.html',
         {'object': task})
+
+
+@login_required
+def select_location(request):
+    next = request.GET.get('next', reverse('campaigns'))
+
+    if request.method == 'POST' and request.POST.get('error') == 'False':
+        request.session['override_location'] = False
```
Comment (smn) [inline ummeli/opportunities/views.py]: This bool to str thing leaves me a bit uncomfortable, please write a test at least?

File: ummeli/opportunities/views.py [hunk 5]
```diff
@@ -58,6 +58,14 @@ def get_queryset(self):
 
         position = self.request.session['location']['position']
 
+        if self.request.session.get('override_location'):
+            tasks = MicroTask.permitted.filter(campaign__pk=campaign.pk)\
+                                        .order_by('province', 'location__city')
+            province = self.request.session['province']
+            if province > 0:
+                tasks = tasks.filter(province=province)
+            return [task for task in tasks if task.available()]
```
Comment (smn) [inline ummeli/opportunities/views.py]: This isn't a queryset. Is it possible to rewrite this to return a query set?

File: ummeli/opportunities/views.py (hunk 3 above)
Comment (smn) [inline ummeli/opportunities/views.py]: Also:

``` python
>>> '' > 0
True
```

File: ummeli/opportunities/urls.py [hunk 6]
```diff
@@ -14,6 +14,9 @@
     url(r'^province(?:/(?P<province>\d+))?/$',
         'ummeli.opportunities.views.change_province',
         name='change_province'),
+    url(r'^microtasks/province(?:/(?P<province>\d+))?/$',
```
Comment (smn) [inline ummeli/opportunities/urls.py]: This regex seems overly complex, can you break down what it does?

Comment (smn) [inline ummeli/opportunities/urls.py]:
``` python
>>> re.match(r'^microtasks/province/(?P<province>\d+)?/?$', 'microtasks/province/').groups()
(None,)
>>> re.match(r'^microtasks/province/(?P<province>\d+)?/?$', 'microtasks/province/12/').groups()
('12',)
>>> 
```

Comment (miltontony) [inline ummeli/opportunities/urls.py]: Interesting. I found that solution as a way to do exactly what you just did. Make the province optional...hmm...I wonder were I initially missed this.

File: ummeli/opportunities/models.py (hunk 2 above)
Comment (miltontony) [inline ummeli/opportunities/models.py:44]: @smn do you think it's OK to raise an exception in that case? I don't expect that to happen, but if it does, it should not be accepted.

File: ummeli/opportunities/views.py (hunk 4 above)
Comment (miltontony) [inline ummeli/opportunities/views.py]: Not to worry, I'll use a form instead. The `clean_data()` should take care of it.

**Comment by smn** [comment]
:+1: :shipit:

**Pull request merged**